R/driver_rds_s3.R

```diff
@@ -0,0 +1,335 @@
+      aws.s3::put_folder(folder = file.path(path, "keys"), bucket = bucket)
+      aws.s3::put_folder(folder = file.path(path, "config"), bucket = bucket)
+      self$bucket <- bucket
+      self$path <- path
+
+      ## This is a bit of complicated dancing around to mantain
+      ## backward compatibility while allowing better defaults in
+      ## future versions.  I'm writing out a version number here that
+      ## future versions of driver_rds can use to patch, warn or
+      ## change behaviour with older versions of the storr.
+      if (!is_new && !s3_object_exists(path = driver_rds_s3_config_file(path, "version"), bucket = bucket)) {
+        s3_write_if_missing("1.0.1", bucket = bucket, path = driver_rds_s3_config_file(path, "version"))
+        s3_write_if_missing("TRUE", bucket = bucket, path =  driver_rds_s3_config_file(path, "mangle_key_pad"))
+        s3_write_if_missing("TRUE", bucket = bucket, path =  driver_rds_s3_config_file(path, "compress"))
+        s3_write_if_missing("md5", bucket = bucket, path =  driver_rds_s3_config_file(path, "hash_algorithm"))
+      }
+      ## Then write out the version number:
+      s3_write_if_missing(as.character(packageVersion("storr")),
+                          bucket = bucket,
+                          path = driver_rds_s3_config_file(path, "version"))
+
+      if (!is.null(mangle_key)) {
+        assert_scalar_logical(mangle_key)
+      }
+      self$mangle_key <- driver_rds_s3_config(bucket, path, "mangle_key", mangle_key,
+                                              FALSE, TRUE)
+
+      if (!is.null(mangle_key_pad)) {
+        assert_scalar_logical(mangle_key_pad)
+      }
+      self$mangle_key_pad <-
+        driver_rds_s3_config(bucket, path, "mangle_key_pad", mangle_key_pad,
+                             FALSE, TRUE)
+
+      if (!is.null(compress)) {
+        assert_scalar_logical(compress)
+      }
+      self$compress <- driver_rds_s3_config(bucket, path, "compress", compress,
+                                            TRUE, FALSE)
+
+      if (!is.null(hash_algorithm)) {
+        assert_scalar_character(hash_algorithm)
+      }
+      self$hash_algorithm <- driver_rds_s3_config(bucket, path, "hash_algorithm",
+                                                  hash_algorithm, "md5", TRUE)
+    },
+
+    type = function() {
+      "rds_s3"
+    },
+    destroy = function() {
+      s3_delete_recursive(bucket = self$bucket, path = self$path)
+    },
+
+    get_hash = function(key, namespace) {
+      s3_readLines(path = self$name_key(key, namespace), bucket = self$bucket)
+    },
+    set_hash = function(key, namespace, hash) {
+      dir_create(self$name_key("", namespace))
+      s3_writeLines(text = hash, path = self$name_key(key, namespace), bucket = self$bucket) #*** should be making use of (or making an equivalent version of) the write_lines function within the storr package here (I think it deletes file if the write fails)
```

Comment:
I agree with the comment - the storr pattern turned up as a way of avoiding corruption and has been useful